machine_learning/fp_growth.py

```diff
@@ -0,0 +1,335 @@
+"""
+The FP-Growth (Frequent Pattern Growth) algorithm is a widely used
+data mining technique for discovering frequent itemsets in
+large transaction databases.
+It overcomes some of the limitations of traditional methods like
+Apriori by efficiently constructing the FP-Tree
+
+WIKI: https://athena.ecs.csus.edu/~mei/associationcw/FpGrowth.html
+Examples: https://www.javatpoint.com/fp-growth-algorithm-in-data-mining
+"""
+
+from typing import Optional
+
+
+class TreeNode:
+    """
+    Initialize a TreeNode.
+
+    Args:
+        name_value (str): The name of the node.
+        num_occur (int): The number of occurrences of the node.
+        parent_node (TreeNode): The parent node.
+
+    Example:
+    >>> parent = TreeNode("Parent", 1, None)
+    >>> child = TreeNode("Child", 2, parent)
+    >>> child.name
+    'Child'
+    >>> child.count
+    2
+    """
+
+    def __init__(
+        self, name_value: str, num_occur: int, parent_node: Optional["TreeNode"] = None
+    ) -> None:
+        self.name = name_value
+        self.count = num_occur
+        self.node_link = None  # Initialize node_link to None
```

Comment:
Add a Python type hint...
```suggestion
        self.node_link: TreeNode | None = None  # Initialize node_link to None
```